### PR TITLE
export network

### PR DESCRIPTION
**Pull request by AbdelHedhili**
Signed-off-by: AbdelHedhili <abdelsalem.hedhili@rte-france.com>

**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
feature

**What is the current behavior?** *(You can also link to an open issue here)*
no export available

**What is the new behavior (if this is a feature change)?**
export of the network linked to a study possible

### CODE REVIEW

File: src/main/java/com/powsybl/network/conversion/server/NetworkConversionService.java [hunk 1]
```diff
@@ -54,6 +60,33 @@ NetworkInfos importCase(UUID caseUuid) {
...
+        MemDataSource memDataSource = new MemDataSource();
+        Network network = networkStoreService.getNetwork(networkUuid);
+        Exporters.export(format, network, null, memDataSource);
+        String extension = getFormatExtension(format);
+        byte[] networkData = memDataSource.getData(extension);
+        String networkName = network.getNameOrId() + extension;
+        return new ExportNetworkInfos(networkName, networkData);
+    }
+
+    String getFormatExtension(String format) {
```
Comment (geofjamg) [inline src/main/java/com/powsybl/network/conversion/server/NetworkConversionService.java:76]: This is breaking the plugin architecture of exporters as we should be able to add a new exporter supporter by just by adding it as a dependency in pom.xml.

The export/datasource design is working like this
  - you give a base name to the exporter
  - the exporter is responsible to add the correct extension to each exported file
  - one export could generated SEVERAL files with different extensions (see XIIDM export with .xml + .csv for id dict)

There is an ongoing work to refactor datasources with a new design, but in the meantime we have to support this one.

I didn't had a look at client side, but are you supporting multiple files export?

Using MemDataSource I guess that you can get file names using 
```java
dataSource.listNames(".*)
```

Comment (AbdelHedhili) [reply src/main/java/com/powsybl/network/conversion/server/NetworkConversionService.java:76]: No I didn't think about multi-files export. I'll correct that.

File: src/main/java/com/powsybl/network/conversion/server/NetworkConversionController.java [hunk 2]
```diff
@@ -43,4 +43,21 @@
         NetworkInfos networkInfos = networkConversionService.importCase(caseUuid);
         return ResponseEntity.ok().body(networkInfos);
     }
+
+    @GetMapping(value = "/networks/{networkUuid}/{format}")
+    @ApiOperation(value = "Export a network from the network-store")
+    public ResponseEntity<ExportNetworkInfos> exportNetwork(@PathVariable("networkUuid") UUID networkUuid, @PathVariable("format") String format) {
```
Comment (jonenst) [inline src/main/java/com/powsybl/network/conversion/server/NetworkConversionController.java:49]: I think we should avoid sending the data (tens of megabytes) as base64 in a json object

File: src/main/java/com/powsybl/network/conversion/server/NetworkConversionException.java [hunk 3]
```diff
@@ -1,11 +1,36 @@
...
+import java.util.Objects;
+
 /**
  * @author Abdelsalem Hedhili <abdelsalem.hedhili at rte-france.com>
  */
-class NetworkConversionException extends RuntimeException {
+public final class NetworkConversionException extends RuntimeException {
+
+    public enum Type {
+        UNSUPPORTED_FORMAT
```
Comment (ChamseddineBhd) [inline src/main/java/com/powsybl/network/conversion/server/NetworkConversionException.java:17]: why an enum that has only one value ? if there's not other NetworkConversionException types we don't need to create an enum !

File: src/main/java/com/powsybl/network/conversion/server/NetworkConversionException.java [hunk 4]
```diff
@@ -1,11 +1,36 @@
...
+    private final Type type;
 
-    NetworkConversionException(String message, Throwable cause) {
-        super(message, cause);
+    private NetworkConversionException(Type type, String msg) {
+        super(msg);
+        this.type = Objects.requireNonNull(type);
     }
+
+    public Type getType() {
```
Comment (ChamseddineBhd) [inline src/main/java/com/powsybl/network/conversion/server/NetworkConversionException.java:27]: same, no need for get type if the class handles one type

File: src/main/java/com/powsybl/network/conversion/server/NetworkConversionException.java [hunk 5]
```diff
@@ -1,11 +1,36 @@
...
+
+    public enum Type {
+        UNSUPPORTED_FORMAT
+    }
+
+    private final Type type;
 
-    NetworkConversionException(String message, Throwable cause) {
-        super(message, cause);
+    private NetworkConversionException(Type type, String msg) {
```
Comment (ChamseddineBhd) [inline src/main/java/com/powsybl/network/conversion/server/NetworkConversionException.java:22]: same no need for type if there's one type

File: src/main/java/com/powsybl/network/conversion/server/NetworkConversionService.java [hunk 6]
```diff
@@ -54,6 +64,44 @@ NetworkInfos importCase(UUID caseUuid) {
...
+
+        Set<String> listNames = memDataSource.listNames(".*");
+        String networkName;
+        byte[] networkData;
+        if (listNames.size() == 1) {
+            networkName = network.getNameOrId() + listNames.toArray()[0];
+            networkData = memDataSource.getData(listNames.toArray()[0].toString());
+        } else {
+            networkName = network.getNameOrId() + ".zip";
+            networkData = createZipFile(listNames.toArray(new String[0]), memDataSource).toByteArray();
```
Comment (ChamseddineBhd) [inline src/main/java/com/powsybl/network/conversion/server/NetworkConversionService.java:83]: i didn't understand that instruction : listNames.toArray(new String[0]) ! why new String[0]

File: src/main/java/com/powsybl/network/conversion/server/NetworkConversionException.java (hunk 3 above)
Comment (AbdelHedhili) [reply src/main/java/com/powsybl/network/conversion/server/NetworkConversionException.java:17]: not yet. but there will be and it'll be easier to add them.

File: src/main/java/com/powsybl/network/conversion/server/NetworkConversionService.java (hunk 6 above)
Comment (AbdelHedhili) [reply src/main/java/com/powsybl/network/conversion/server/NetworkConversionService.java:83]: to convert the list to a String array

File: src/main/java/com/powsybl/network/conversion/server/NetworkConversionException.java (hunk 4 above)
Comment (AbdelHedhili) [reply src/main/java/com/powsybl/network/conversion/server/NetworkConversionException.java:27]: the goal is to be extensible. That's basically the way we did it in all our services

Comment (AbdelHedhili) [reply src/main/java/com/powsybl/network/conversion/server/NetworkConversionException.java:27]: The goal is to be extensible. That's basically the way we did it in all our services

File: src/main/java/com/powsybl/network/conversion/server/NetworkConversionException.java (hunk 5 above)
Comment (AbdelHedhili) [reply src/main/java/com/powsybl/network/conversion/server/NetworkConversionException.java:22]: same reason

File: src/main/java/com/powsybl/network/conversion/server/NetworkConversionService.java [hunk 7]
```diff
@@ -36,6 +44,8 @@
     @Autowired
     private NetworkStoreService networkStoreService;
 
+    private static final Logger LOGGER = LoggerFactory.getLogger(NetworkConversionService.class);
```
Comment (ChamseddineBhd) [inline src/main/java/com/powsybl/network/conversion/server/NetworkConversionService.java:47]: please remove that unused Logger

Comment (AbdelHedhili) [reply src/main/java/com/powsybl/network/conversion/server/NetworkConversionService.java]: done

File: src/main/java/com/powsybl/network/conversion/server/NetworkConversionController.java [hunk 8]
```diff
@@ -43,4 +45,22 @@
         NetworkInfos networkInfos = networkConversionService.importCase(caseUuid);
         return ResponseEntity.ok().body(networkInfos);
     }
+
+    @GetMapping(value = "/networks/{networkUuid}/{format}")
```
Comment (geofjamg) [inline src/main/java/com/powsybl/network/conversion/server/NetworkConversionController.java:49]: /networks/{networkUuid}/export/{format} to be consistent with /export/formats ?

**Pull request merged**